### PR TITLE
HA tests - Replication 1/2

### PR DESCRIPTION
**Pull request by paulomach**
# Issue

All charms must be CI tested for Replication/HA scenarios.
Jira tasks [DPE424](https://warthogs.atlassian.net/browse/DPE-424), [DPE425](https://warthogs.atlassian.net/browse/DPE-425), [DPE426](https://warthogs.atlassian.net/browse/DPE-426), [DPE427](https://warthogs.atlassian.net/browse/DPE-427)

Self Healing (2/2) test in the follow up PR.

# Solution

Implement missing HA test (cluster isolation) plus support for running against a existing juju deployed cluster (through ` -- -model <model-name>` pos-args)

# Testing

New Integration tests implemented.

# Release Notes

Important to review:
* `tests/integration/helpers.py` - couple new methods
* `tests/integration/test_ha.py` - new tests on renamed file

Not important:
* `tests/unit/test_relation_mysql_legacy.py` - removed workaround due bug resolution in ops-1.5.2
* `tests/unit/test_shared_db.py` - same as above

### CODE REVIEW

File: tests/integration/test_ha.py [hunk 1]
```diff
@@ -243,3 +260,88 @@ async def test_cluster_preserves_data_on_delete(ops_test: OpsTest) -> None:
...
+
+    This test creates a new cluster, create a new table on both cluster, write a single record with
+    the application name for each cluster, retrieve and compare these records, asserting they are
+    not the same.
+    """
+    app = await app_name(ops_test)
+    apps = [app, ANOTHER_APP_NAME]
+
+    # Build and deploy secondary charm
+    charm = another_charm or await ops_test.build_charm(".")
```
Comment (zmraul) [inline tests/integration/test_ha.py:278]: Is there a situation where you need to build here again, instead of just using `another_charm`?

Comment (zmraul) [inline tests/integration/test_ha.py:278]: Is there a situation where you need to build here again, instead of just using `another_charm`?

**Review by zmraul — APPROVED** [review]
This is a really nice test!

File: tests/integration/test_ha.py (hunk 1 above)
Comment (paulomach) [reply tests/integration/test_ha.py:278]: Yes, when running against a previously running model - the charm file could even be already built, but there's no reference to it.

**Review by WRFitch — APPROVED** [review]
Nice

**Pull request merged**